### PR TITLE
Fix copy local files to remote target folder with sub-folder

### LINKED ISSUE #6767: CalcJob bug in coping local files to remote folder with sub-folder
I got an error when running [aiida-hubbard](https://github.com/aiidateam/aiida-hubbard) using `aiida-core>=2.6.0`,

In the HP `CalcJob`, its `local_copy_list` is:
```
local_copy_list [(uuid 'out/HP', 'out/HP')]
```

The expected_hierarchy should be `out/HP/the-files`, but I got `out/HP/HP/the-files`. This does not happen for `aiida-core < 2.6`.

As suggested by @bastonero , this could be introduced in this [PR](https://github.com/aiidateam/aiida-core/commit/6898ff4d8c263cf08707c61411a005f6a7f731dd).

### PR DESCRIPTION
**Pull request by superstar54**
Fix #6767 .

### CODE REVIEW

File: src/aiida/engine/daemon/execmanager.py [hunk 1]
```diff
@@ -373,7 +373,7 @@ async def _copy_local_files(logger, node, transport, inputs, local_copy_list, wo
                 # If the source object is a directory, we copy its entire contents
                 data_node.base.repository.copy_tree(filepath_target, filename_source)
                 await transport.put_async(
-                    f'{dirpath}/*',
```
Comment (superstar54) [inline src/aiida/engine/daemon/execmanager.py:376]: This will copy everything inside `dirpath` to `workdir`. If `filename_target` is a `out/HP`, then the `dirpath` will have `out/HP`, then it will copy them into the workdir target `out/HP`, result folders as reported in #6767

**Comment by superstar54** [comment]
Pining @bastonero , so he is aware of this fix.

**Comment by bastonero** [comment]
Thanks @superstar54 ! As soon as this is released in a newer version i will update the deps of aiida-hubbard so to support the latest aiida-core version. Thanks a lot again for this hot fix!

File: src/aiida/transports/plugins/ssh.py [hunk 2]
```diff
@@ -729,6 +729,12 @@ def mkdir(self, path: TransportPath, ignore_existing: bool = False):
         if ignore_existing and self.isdir(path):
             return
 
+        # check and create the parent directories recursively
+        parent_path = os.path.dirname(path)
+
+        if not self.isdir(parent_path):
+            self.mkdir(parent_path, ignore_existing=True)
```
Comment (khsrali) [inline src/aiida/transports/plugins/ssh.py:736]: Also node we already have this functionality in the same file, search for `def makedirs`

**Review by khsrali — CHANGES_REQUESTED** [review]
Thanks @superstar54 for your suggestion.
I think this was rahter introduced along with https://github.com/aiidateam/aiida-core/pull/6447

And this fix can potentially, mix up things. 

I ping @mbercx , if he can approve these changes. As he originally wrote that part of the code, so he's the master of it :wink:

**Comment by mbercx** [comment]
Thanks for the ping @khsrali! I remember that indeed some of the behaviour of the copy lists was somewhat counter-intuitive, but changing it could potentially break quite a few plugins, so we need to take care. See this PR for a lot more context and discussion:

https://github.com/aiidateam/aiida-core/pull/6348

I would need some time to work back into this to do a proper review, and that's going to be tricky this week for sure.

**Comment by unkcpz** [comment]
Since the same problem show up again and again and the legacy code can take quite some effort to understand to and review. I think may be we can go back to the source of the problem and make some improvement to the final API that is exposed to the user. 

From the OP posted by @superstar54 the problem is `local_copy_list` call is not expressive enough to distinguish the folder and the files. Even the user know what they are doing, there is no clear way how they can explicitly tell the function call it is a folder and he want it not in a nested folder. I think the original  design is try to mimic what Linux is doing for the file copy but even different linux posix has different standard. Then why not forming our own API and make the `local_copy_list` call more explicit?  

@khsrali @mbercx since you are experts on tranport and remote file modification components, I looking forward for your opinion.

**Comment by khsrali** [comment]
@unkcpz thanks for your thoughts.

I think `local_copy_list` is considered to be a public api, since plugin developers have to fill it up with `CodeInfo().local_copy_list` property.  It appears to be a list of tuple of  `(uuid, dest_path, rel_path)`.

I assume the source of problem, is the complexity of the whole sandbox logic, that it used to do inconstant behavior after fixing a bug: https://github.com/aiidateam/aiida-core/pull/6348

However, @mbercx fix what broke at https://github.com/aiidateam/aiida-core/pull/6447.

Now it could be, that those series of fixes, somehow missed a marginal case that broke the backward compatibility.
I don't know what solution to provide. One has to look behind and re-understand what was broke and what was the fix.

**Comment by superstar54** [comment]
Hi @khsrali , @mbercx , @unkcpz , could we fix a meeting to discuss this PR?

**Comment by mbercx** [comment]
Thanks for the ping @superstar54, had to work myself into this a bit again. 

## Sanity check

First I wanted to sanity check that this PR reproduces the behaviour of v2.5.1, and that it was subsequently broken in v2.6.3. For this I added a test case to the `test_upload_combination` representing https://github.com/aiidateam/aiida-core/issues/6767, mainly because I still had an old branch for v2.5.1 where I added these tests when working on https://github.com/aiidateam/aiida-core/pull/6447. Note that I can't introduce the test you add here easily for v2.5.1, because the various copy commands hadn't been separated yet.

I can indeed confirm that:

1. The test passes for v2.5.1 ([branch](https://github.com/mbercx/aiida-core/tree/test/v2.5.1), [tests](https://github.com/mbercx/aiida-core/actions/runs/14370880405/job/40293582005)).
2. It fails for v2.6.3 ([branch](https://github.com/mbercx/aiida-core/tree/test/v2.6.3), [tests](https://github.com/mbercx/aiida-core/actions/runs/14371019149/job/40293969596)). Note that there are two failures: for the `core.local` transport the hierarchy is incorrect, as you note above. For the `core.ssh` one, the `mkdir` command fails, which is why indeed you made changes there (more comments on that below).
3. Your changes here fix the issue, reproducing the behaviour of v2.5.1 ([branch](https://github.com/mbercx/aiida-core/tree/pr/superstar54/6768), [tests](https://github.com/mbercx/aiida-core/actions/runs/14370993041/job/40293887767)).

So, I'm satisfied that your fix is a good one. We clearly missed this edge case previously.

## Behaviour of `Transport.puttree()`

The second issue you are addressing here is a discrepancy in the behaviour of `LocalTransport.puttree()` and `SshTransport.puttree()`.
Say we have the following contents of an `out` directory:

```console
❯ tree out
out
└── HP
    └── file
```

Similar to the `hp.x` example from the issue (`local_copy_list [(uuid 'out/HP', 'out/HP')]`).
If I use the `LocalTransport.puttree()` command, this works perfectly fine:

```python
with orm.load_computer('localhost').get_transport() as transport:
    transport.puttree(
        (Path('/Users/mbercx/project/core/tmp') / 'out' / 'HP').as_posix(),
        (Path().cwd() / 'out' / 'HP').as_posix(),
    )
```

The command executes without issue, and the expected file hierarchy is reproduced:

```
out
└── HP
    └── file
```

However, this command fails for the `SshTransport`, in case the `out` directory doesn't already exist on the target:

```python
OSError: Error during mkdir of '/users/mabercx/tmp/out/HP', maybe you don't have the permissions to do it, or the directory already exists? ([Errno 2] No such file)
```

Which is why you made the changes to `SshTransport.mkdir()`.
So we aren't just correcting the behaviour of `remote_copy_list` here, we're also changing how `SshTransport.puttree()` works.

## My suggestions

As @khsrali suggested, I would use the `makedirs` method for both SSH transports to have align their behaviour with the local transport, see https://github.com/mbercx/aiida-core/commit/397d22d11a04064c93137c292b9651b44c08eb95.

I'm happy to add your fix here for the `_copy_local_files`, and added the HP test case so far here: https://github.com/mbercx/aiida-core/commit/5b2505ae09b90f84d9026a676811e53f039f3c94.
We could still look for more edge cases and see if we can make sure the behavior is the same for v2.5.1 and the next release, but I don't think we'll ever catch them all (there are so many options...).
Since v2.6.3 is out for a while, let's hope we didn't break any other behavior and we're in good shape with this fix.

In general, I agree with @unkcpz that perhaps the copy lists API needs an overhaul, and we should also have a careful look at the various transports and make sure the behaviour of their commands correspond. But since this fix is critical for the release, I would go ahead for now. Redesigning the copy lists etc is a big change that would most likely be backwards incompatible.

**Comment by superstar54** [comment]
Hi @mbercx , thanks for the detailed checking. Please go ahead with your fix.

**Comment by mbercx** [comment]
Thank _you_ for picking up the slack I left behind 😉 . Can I force-push to this branch?

@khsrali did you still want to have a look at the transports fix?

**Comment by superstar54** [comment]
> Can I force-push to this branch?

Sure, please go ahead.

**Comment by khsrali** [comment]
Cheers, @mbercx! thanks a lot for writing a new test, please ping me as well once ready for review

> @khsrali did you still want to have a look at the transports fix?

I think make sense to add a fix for `LocalTransport.mkdir()` to raise if a tree is passed. So to unify the behavior with `core.ssh`

**Comment by mbercx** [comment]
> I think make sense to add a fix for LocalTransport.mkdir() to raise if a tree is passed. So to unify the behavior with core.ssh

I'm afraid I don't understand this suggestion. 😅 Trying to run `LocalTransport.mkdir()` with a nested directory whose parent doesn't exist already fails:

```python
with orm.load_computer('localhost').get_transport() as transport:
    transport.mkdir(Path().cwd() / 'new' / 'nested')
```

```python
---------------------------------------------------------------------------
FileNotFoundError                         Traceback (most recent call last)
Cell In[12], [line 2](vscode-notebook-cell:?execution_count=12&line=2)
      [1](vscode-notebook-cell:?execution_count=12&line=1) with orm.load_computer('localhost').get_transport() as transport:
----> [2](vscode-notebook-cell:?execution_count=12&line=2)     transport.mkdir(Path().cwd() / 'new' / 'nested')

File ~/project/core/git/aiida-core/src/aiida/transports/plugins/local.py:196, in LocalTransport.mkdir(self, path, ignore_existing)
    [193](https://file+.vscode-resource.vscode-cdn.net/Users/mbercx/project/core/jupyter/~/project/core/git/aiida-core/src/aiida/transports/plugins/local.py:193) if ignore_existing and self.isdir(path):
    [194](https://file+.vscode-resource.vscode-cdn.net/Users/mbercx/project/core/jupyter/~/project/core/git/aiida-core/src/aiida/transports/plugins/local.py:194)     return
--> [196](https://file+.vscode-resource.vscode-cdn.net/Users/mbercx/project/core/jupyter/~/project/core/git/aiida-core/src/aiida/transports/plugins/local.py:196) os.mkdir(os.path.join(self.curdir, path))

FileNotFoundError: [Errno 2] No such file or directory: '/Users/mbercx/project/core/jupyter/new/nested'
```

**Comment by mbercx** [comment]
@khsrali ah yes, this is exactly the change I also had in mind (see https://github.com/aiidateam/aiida-core/pull/6768/commits/397d22d11a04064c93137c292b9651b44c08eb95). I've force-pushed my local changes here, splitting them up in two commits. Please use "rebase and merge" when merging after the review, so they are preserved. :)

File: tests/engine/daemon/test_execmanager.py [hunk 3]
```diff
@@ -354,6 +355,16 @@ def test_upload_file_copy_operation_order(node_and_calc_info, tmp_path, order, e
             {'Ba.upf': 'Ba pseudo'},
             None,
         ),
+        # Only local copy of a single nested directory to the same nested directory
+        # -> Copies the contents of the nested folder to the target nested folder
+        # COUNTER-INTUITIVE: this command would fail with `cp` since the parent folder does not exist
+        (
+            {},
+            ((FolderData, {'out': {'HP': {'file': 'content'}}}, 'out/HP', 'out/HP'),),
```
Comment (khsrali) [inline tests/engine/daemon/test_execmanager.py:363]: I'm a bit confused here, `'out/HP'` is both

```
        - The name of the file or directory to be copied.
        - The relative path the data should be copied to.
```
Ok for the first.
But from the second, I expect something like this to happen: 
`out/HP/out/HP/file`  
either the explanation should be improved or why is this working?

**Review by khsrali — COMMENTED** [review]
Thanks a lot @mbercx
Just a confusion that I asked below

File: tests/engine/daemon/test_execmanager.py (hunk 3 above)
Comment (mbercx) [reply tests/engine/daemon/test_execmanager.py:363]: I understand these commands can be confusing, it's one of the reasons I hate coming back to this code. ^^

Which explanation should we improve though? Looking at the documentation of the `local_copy_list`:

https://aiida.readthedocs.io/projects/aiida-core/en/stable/topics/calculations/usage.html#local-copy-list

I think the behavior described here matches exactly that of this test. Moreover, it's the behavior of `local_copy_list` pre-v2.6. 

I would expect `out/HP/out/HP/file` in case the source path is `'.'`, in fact I've added a test to show this case.

Comment (khsrali) [reply tests/engine/daemon/test_execmanager.py:363]: >I understand these commands can be confusing, it's one of the reasons I hate coming back to this code. ^^

Totally resonate with that :smiling_face_with_tear: 
Although everything works, let's fix this documentation and confusion once for ever.
The source of my confusion is that the `local_copy_list` of `calc_info` is not the same as `local_copy_list` parameter in  `test_upload_combinations`

>Which explanation should we improve though? Looking at the documentation of the local_copy_list:
>https://aiida.readthedocs.io/projects/aiida-core/en/stable/topics/calculations/usage.html#local-copy-list

From the link you sent it defines `local_copy_list` as:
```
The local copy list takes tuples of length three, each of which represents a file or directory to be copied, defined through the following items:

node uuid: the node whose repository contains the file, typically a SinglefileData or FolderData node
source relative path: the relative path of the file or directory within the node repository
target relative path: the relative path within the working directory to which to copy the file or directory contents
```

While in the docstring of `def test_upload_combinations` defines `local_copy_list`, a pytest parameter, as:
```
The `local_copy_list` is formatted as a list of tuples, where each tuple contains the following elements:
    - The class of the data node to be copied.
    - The content of the data node to be copied. This can be either a string in case of a file, or a dictionary
        representing the file hierarchy in case of a folder.
    - The name of the file or directory to be copied.
    - The relative path the data should be copied to.
```
[Here](https://github.com/aiidateam/aiida-core/blob/660fec70ef43a64be7edae3c12f0a0fd5ef84349/tests/engine/daemon/test_execmanager.py#L564)  `local_copy_list` parameter in `test_upload_combinations` is transformed into `local_copy_list` property of `calcinfo` :
```python
for copy_id, (data_class, content, filename, target_path) in enumerate(local_copy_list):
    # Create a sub directroy in the temporary folder for each copy to avoid conflicts
    sub_tmp_path_local = tmp_path / f'local_{copy_id}'
...
    elif issubclass(data_class, FolderData):
        create_file_hierarchy(content, sub_tmp_path_local)
        serialize_file_hierarchy(sub_tmp_path_local, read_bytes=False)
        folder = FolderData()
        folder.base.repository.put_object_from_tree(sub_tmp_path_local)
        folder.store()

        calc_info.local_copy_list.append((folder.uuid, filename, target_pat
```

Now, I try to break it down and understand that translation. Your `local_copy_list` parameter is:
```python
((FolderData, {'out': {'HP': {'file': 'content'}}}, 'out/HP', 'out/HP'),),
```

And the test is doing:
```python
calc_info.local_copy_list.append((folder.uuid, 'out/HP', 'out/HP'))
```

Therefore it's in fact copying `'/<src>/out/HP/*'` to `'/<dst>/out/HP/*'`

That took me an hour and a half to understand and formulate this comment. Therefor I'm suggesting the following changes:

1) Refactor that fake `local_copy_list` --pytest parameter-- to something else.
2) change the docstring of `test_upload_combinations as follows`:
```
The `local_copy_list` is formatted as a list of tuples, where each tuple contains the following elements:
    - The class of the data node to be copied.
    - The content of the data node to be copied. This can be either a string in case of a file, or a dictionary
        representing the file hierarchy in case of a folder.
    - The path of the file or directory to be copied, specified relative to the source node's working directory.
    - The destination path where the data will be copied, specified relative to the destination node's working directory.
```
3) change that explanation in the link you send, in the usage section:
```
The local copy list consists of three-element tuples, each defining a file or directory to be copied. The tuple contains the following items:

Node UUID: The UUID of the node (e.g. a SinglefileData or FolderData) containing the file/directory.

Source Relative Path: The path of the file/directory within the node's repository, relative to its base directory.

Target Relative Path: The destination path relative to the working directory of the target node, where the file/directory contents will be copied.
```

Comment (mbercx) [reply tests/engine/daemon/test_execmanager.py:363]: Thanks for the extensive description @khsrali! Happy to make the changes to the `test_upload_combinations` docstring and fine-tune the documentation a bit. Just one point of clarification on your comments:

> Refactor that fake local_copy_list --pytest parameter-- to something else.

You mean give it a different name to avoid confusion?

Comment (khsrali) [reply tests/engine/daemon/test_execmanager.py:363]: > You mean give it a different name to avoid confusion?

yes exactly, something like `local_copy_list_params`

File: src/aiida/transports/plugins/ssh_async.py [hunk 4]
```diff
@@ -572,7 +572,7 @@ async def puttree_async(
             raise OSError('Cannot copy a directory into a file')
 
         if not await self.isdir_async(remotepath):  # in this case copy things in the remotepath directly
-            await self.mkdir_async(remotepath)  # and make a directory at its place
+            await self.makedirs_async(remotepath, ignore_existing=overwrite)  # and make a directory at its place
```
Comment (khsrali) [inline src/aiida/transports/plugins/ssh_async.py:575]:
```suggestion
            await self.makedirs_async(remotepath)  # and make a directory at its place
```
`remotepath` will not ever exist in this case, since the `if` condition is already checking that :)

**Review by khsrali — APPROVED** [review]
Thanks a lot, @mbercx!

I added a minor comment, feel free to ignore.
I just approve, merge whenever/however you want.

File: src/aiida/transports/plugins/ssh_async.py (hunk 4 above)
Comment (mbercx) [reply src/aiida/transports/plugins/ssh_async.py:575]: Fair point! Will adapt :)
